packages/phone-number-privacy/common/src/utils/authentication.ts

```diff
@@ -0,0 +1,73 @@
+import { ContractKit } from '@celo/contractkit'
+import { AuthenticationMethod } from '@celo/contractkit/lib/identity/odis/query'
+import { AccountsWrapper } from '@celo/contractkit/lib/wrappers/Accounts'
+import { trimLeading0x } from '@celo/utils/lib/address'
+import { retryAsyncWithBackOff } from '@celo/utils/lib/async'
+import { verifySignature } from '@celo/utils/lib/signatureUtils'
+import { ec as EC } from 'elliptic'
+import { Request } from 'firebase-functions'
+import { RETRY_COUNT, RETRY_DELAY_IN_MS } from './constants'
+import logger from './logger'
+
+const ec = new EC('secp256k1')
+
+/*
+ * Confirms that user is who they say they are and throws error on failure to confirm.
+ * Authorization header should contain the EC signed body
+ */
+export async function authenticateUser(
+  request: Request,
+  contractKit: ContractKit
+): Promise<boolean> {
+  logger.debug('Authenticating user')
+
+  // https://tools.ietf.org/html/rfc7235#section-4.2
+  const messageSignature = request.get('Authorization')
+  const message = JSON.stringify(request.body)
+  const signer = request.body.account
+  const authMethod = request.body.authenticationMethod
+
+  if (!messageSignature || !signer) {
+    return false
+  }
+
+  if (authMethod && authMethod === AuthenticationMethod.ENCRYPTION_KEY) {
+    const registeredEncryptionKey = await getDataEncryptionKey(signer, contractKit)
+    if (!registeredEncryptionKey) {
+      logger.info(`Account ${signer} does not have registered encryption key`)
+    } else {
+      logger.info(`Found DEK ${registeredEncryptionKey} for ${signer}`)
+      try {
+        const key = ec.keyFromPublic(trimLeading0x(registeredEncryptionKey), 'hex')
+        const parsedSig = JSON.parse(messageSignature)
```

Comment:
Why do we JSON parse here?

Reply:
Chatted offline. It looks like the sender is serializing a DER object but it's not clear whether it's a string or not. We'll leave as-is for now since we'll be migrating this at some point to EIP712 standard.